src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
         return ref.get().getPendingRepair();
     }
 
-    public int getManifestSize()
-    {
-        return manifest.components().size();
-    }
-
     @Override
     public void write(StreamSession session, DataOutputStreamPlus out, int version) throws IOException
     {
+        boolean streamEntirely = shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus;
         SSTableReader sstable = ref.get();
-        CassandraStreamHeader.serializer.serialize(header, out, version);
-        out.flush();
 
-        if (shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus)
+        if (streamEntirely)
         {
-            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
-            writer.write((AsyncStreamingOutputPlus) out);
+            // Acquire lock to avoid concurrent sstable component mutation because of stats update or index summary
+            // redistribution, otherwise file sizes recorded in component manifest will be different from actual
+            // file sizes. (Note: Windows doesn't support atomic replace and index summary redistribution deletes
+            // existing file first)
+            // Recreate the latest manifest and hard links for mutatable components in case they are modified.
+            try (ComponentContext context = sstable.runWithLock(ignored -> ComponentContext.create(sstable.descriptor)))
+            {
+                CassandraStreamHeader current = makeHeader(sstable, operation, sections, estimatedKeys, true, context.manifest());
+                CassandraStreamHeader.serializer.serialize(current, out, version);
+                out.flush();
+
+                CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, context);
+                writer.write((AsyncStreamingOutputPlus) out);
+            }
         }
         else
         {
+            // legacy streaming is not affected by stats metadata mutation and index sumary redistribution
+            CassandraStreamHeader.serializer.serialize(header, out, version);
+            out.flush();
+
             CassandraStreamWriter writer = (header.compressionInfo == null) ?
```

Comment:
The code around `CassandraStreamHeader.compressionInfo` is really misleading. On the sender side `compressinInfo` will be `null` until the header is serialized (the serialization calling `CassandraStreamHeader.calculateCompressionInfo`). Until that point the `CassandraStreamHeader.isCompressed` will return `false` even if the stream should be compressed. `equals` and `hashcode` are also affected by this effect.
For `CassandraStreamHeader.isCompressed` should check if the `compressionMetadata` is different from `null` and `compressionInfo`can be removed from `equals` and `hashcode` as they already use `compressionMetadata` and `sections` in their computation.
Regarding the `compressionInfo` field I wonder if it will not be better to simply make it as private and provide an accessor method that instantiate it lazily. It will be slightly less efficient but probably much safer. What do you think?

Reply:
> CassandraStreamHeader.isCompressed will return false even if the stream should be compressed.

`isCompressed()` is used on received side where `compressionMetadata` is always `null` regardless compressed. But good to make it consistent.

> Regarding the compressionInfo field I wonder if it will not be better to simply make it as private and provide an accessor method that instantiate it lazily.

Good idea, added accessor `getOrInitCompressionInfo`

> compressionInfo can be removed from equals and hashcode as they already use compressionMetadata and sections in their computation.

`equals`/`hashCode` didn't use `compressionMetadata`. Since `equals/hashCode` is not used in production code, I have changed them to use `getOrInitCompressionInfo()`, wdyt?

Reply:
I missed the receiving side logic with `compressionMetadata` vs `compressionInfo`. The changes you did will work, I just wonder if we cannot make the logic even less confusing. I tried to refactor things a bit and came up with:
https://github.com/apache/cassandra/commit/5c01763f7f8ae6b58a2d795101465d8195bf56d4
Tell me if it makes sense to you.

Reply:
Thanks for the refactoring, it looks great. I made a minor fix to use `Arrays.equals` for `chunks()` in `CompressionInfo#equals` https://github.com/apache/cassandra/pull/642/commits/7522f57e9397d94820a1020c1bba808b8a262e1a

Reply:
Good catch :-)